File: test/config/router/route_guards_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:maki_mobile_pos/config/router/route_guards.dart';
import 'package:maki_mobile_pos/config/router/route_names.dart';
import 'package:maki_mobile_pos/core/enums/user_role.dart';
import 'package:maki_mobile_pos/domain/entities/user_entity.dart';

UserEntity _user(UserRole role, {bool isActive = true}) => UserEntity(
      id: 'u-${role.value}',
      email: '${role.value}@test',
      displayName: '${role.value} user',
      role: role,
      isActive: isActive,
      createdAt: DateTime(2025, 1, 1),
    );

void main() {
  group('RouteGuards.isPublicRoute', () {
    test('login is public', () {
      expect(RouteGuards.isPublicRoute(RoutePaths.login), true);
    });

    test('any authenticated path is not public', () {
      expect(RouteGuards.isPublicRoute(RoutePaths.dashboard), false);
      expect(RouteGuards.isPublicRoute(RoutePaths.pos), false);
      expect(RouteGuards.isPublicRoute(RoutePaths.users), false);
    });
  });

  group('RouteGuards.canAccess — unauthenticated', () {
    test('null user can access /login', () {
      expect(RouteGuards.canAccess(RoutePaths.login, null), true);
    });

    test('null user cannot access any protected route', () {
      expect(RouteGuards.canAccess(RoutePaths.dashboard, null), false);
      expect(RouteGuards.canAccess(RoutePaths.pos, null), false);
      expect(RouteGuards.canAccess(RoutePaths.inventory, null), false);
      expect(RouteGuards.canAccess(RoutePaths.users, null), false);
    });
  });

  group('RouteGuards.canAccess — inactive user', () {
    test('inactive admin is denied every protected route', () {
      final inactive = _user(UserRole.admin, isActive: false);
      expect(RouteGuards.canAccess(RoutePaths.dashboard, inactive), false);
      expect(RouteGuards.canAccess(RoutePaths.pos, inactive), false);
      expect(RouteGuards.canAccess(RoutePaths.inventory, inactive), false);
      expect(RouteGuards.canAccess(RoutePaths.users, inactive), false);
    });

    test('inactive user can still access /login (public)', () {
      final inactive = _user(UserRole.cashier, isActive: false);
      expect(RouteGuards.canAccess(RoutePaths.login, inactive), true);
    });
  });

  // The matrix: row = role, columns = routes. Entries are the expected
  // canAccess() outcome. Common routes (POS, drafts, dashboard) should be true
  // for every role; admin-only routes should be true only for admin.
  group('RouteGuards.canAccess — role × route matrix', () {
    final cashier = _user(UserRole.cashier);
    final staff = _user(UserRole.staff);
    final admin = _user(UserRole.admin);

    final matrix = <String, _Expected>{
      RoutePaths.dashboard:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.pos: const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.checkout:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.drafts:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.inventory:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.productAdd:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.receiving:
          const _Expected(cashier: false, staff: true, admin: true),
      RoutePaths.bulkReceiving:
          const _Expected(cashier: false, staff: true, admin: true),
      RoutePaths.suppliers:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.supplierAdd:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.expenses:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.expenseAdd:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.reports:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.salesReport:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.profitReport:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.users:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.userAdd:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.settings:
          const _Expected(cashier: true, staff: true, admin: true),
      RoutePaths.costCodeSettings:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.userLogs:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.pettyCash:
          const _Expected(cashier: false, staff: false, admin: true),
      RoutePaths.pettyCashNew:
          const _Expected(cashier: false, staff: false, admin: true),
    };

    matrix.forEach((path, expected) {
      test('$path  cashier=${expected.cashier}, staff=${expected.staff}, admin=${expected.admin}',
          () {
        expect(RouteGuards.canAccess(path, cashier), expected.cashier,
            reason: 'cashier should ${expected.cashier ? "" : "not "}access $path');
        expect(RouteGuards.canAccess(path, staff), expected.staff,
            reason: 'staff should ${expected.staff ? "" : "not "}access $path');
        expect(RouteGuards.canAccess(path, admin), expected.admin,
            reason: 'admin should ${expected.admin ? "" : "not "}access $path');
      });
    });
  });

  group('RouteGuards.canAccess — dynamic routes', () {
    test('inventory edit allows staff (limited) and admin (full)', () {
      expect(
          RouteGuards.canAccess(
              '/inventory/edit/abc123', _user(UserRole.staff)),
          true);
      expect(
          RouteGuards.canAccess(
              '/inventory/edit/abc123', _user(UserRole.admin)),
          true);
    });

    test('inventory edit denies cashier (no edit permission)', () {
      expect(
          RouteGuards.canAccess(
              '/inventory/edit/abc123', _user(UserRole.cashier)),
          false);
    });

    test('expense edit denies cashier and staff (admin-only)', () {
      expect(
          RouteGuards.canAccess(
              '/expenses/edit/exp-1', _user(UserRole.cashier)),
          false);
      expect(
          RouteGuards.canAccess(
              '/expenses/edit/exp-1', _user(UserRole.staff)),
          false);
      expect(
          RouteGuards.canAccess(
              '/expenses/edit/exp-1', _user(UserRole.admin)),
          true);
    });

    test('user edit is admin-only', () {
      expect(
          RouteGuards.canAccess(
              '/users/edit/u-1', _user(UserRole.cashier)),
          false);
      expect(
          RouteGuards.canAccess(
              '/users/edit/u-1', _user(UserRole.staff)),
          false);
      expect(RouteGuards.canAccess('/users/edit/u-1', _user(UserRole.admin)),
          true);
    });

    test('supplier edit is admin-only', () {
      expect(
          RouteGuards.canAccess(
              '/suppliers/edit/s-1', _user(UserRole.staff)),
          false);
      expect(
          RouteGuards.canAccess(
              '/suppliers/edit/s-1', _user(UserRole.admin)),
          true);
    });

    test('inventory detail (view) is allowed for any role with viewInventory',
        () {
      expect(
          RouteGuards.canAccess('/inventory/p-1', _user(UserRole.cashier)),
          true);
      expect(RouteGuards.canAccess('/inventory/p-1', _user(UserRole.staff)),
          true);
      expect(RouteGuards.canAccess('/inventory/p-1', _user(UserRole.admin)),
          true);
    });

    test('draft edit (e.g. /drafts/abc) is treated as common route', () {
      expect(RouteGuards.canAccess('/drafts/abc', _user(UserRole.cashier)),
          true);
      expect(RouteGuards.canAccess('/drafts/abc', _user(UserRole.staff)),
          true);
      expect(RouteGuards.canAccess('/drafts/abc', _user(UserRole.admin)),
          true);
    });
  });

  group('RouteGuards.getMenuItems', () {
    test('cashier sees POS, Drafts, Inventory, Expenses, Reports, Settings',
        () {
      final paths = RouteGuards.getMenuItems(UserRole.cashier)
          .map((e) => e.path)
          .toSet();
      expect(paths, contains(RoutePaths.pos));
      expect(paths, contains(RoutePaths.drafts));
      expect(paths, contains(RoutePaths.inventory));
      expect(paths, contains(RoutePaths.expenses));
      expect(paths, contains(RoutePaths.reports));
      expect(paths, contains(RoutePaths.settings));
      expect(paths, isNot(contains(RoutePaths.receiving)));
      expect(paths, isNot(contains(RoutePaths.suppliers)));
      expect(paths, isNot(contains(RoutePaths.users)));
      expect(paths, isNot(contains(RoutePaths.pettyCash)));
      expect(paths, isNot(contains(RoutePaths.userLogs)));
    });

    test('staff additionally sees Receiving', () {
      final paths = RouteGuards.getMenuItems(UserRole.staff)
          .map((e) => e.path)
          .toSet();
      expect(paths, contains(RoutePaths.receiving));
      expect(paths, isNot(contains(RoutePaths.suppliers)));
      expect(paths, isNot(contains(RoutePaths.users)));
      expect(paths, isNot(contains(RoutePaths.pettyCash)));
      expect(paths, isNot(contains(RoutePaths.userLogs)));
    });

    test('admin sees the full menu including Suppliers, Users, Logs, Petty Cash',
        () {
      final paths = RouteGuards.getMenuItems(UserRole.admin)
          .map((e) => e.path)
          .toSet();
      expect(paths, contains(RoutePaths.suppliers));
      expect(paths, contains(RoutePaths.users));
      expect(paths, contains(RoutePaths.userLogs));
      expect(paths, contains(RoutePaths.pettyCash));
    });
  });
}

class _Expected {
  final bool cashier;
  final bool staff;
  final bool admin;
  const _Expected({
    required this.cashier,
    required this.staff,
    required this.admin,
  });
}

File: web_admin/src/presentation/router/routeGuards.test.ts
import { describe, expect, it } from 'vitest';
import { canAccess } from './routeGuards';
import { RoutePaths } from './routePaths';
import { UserRole } from '@/domain/enums';
import type { User } from '@/domain/entities/User';

const admin: User = {
  id: 'u1',
  email: 'admin@shop.test',
  displayName: 'Admin',
  role: UserRole.admin,
  isActive: true,
  phoneNumber: null,
  photoUrl: null,
  createdAt: new Date('2026-01-01'),
  updatedAt: null,
  createdBy: null,
  updatedBy: null,
  lastLoginAt: null,
};

const cashier: User = { ...admin, id: 'u2', email: 'cashier@shop.test', role: UserRole.cashier };
const staff: User = { ...admin, id: 'u3', email: 'staff@shop.test', role: UserRole.staff };

describe('canAccess — settings routes', () => {
  it('admin can reach the admin-managed list pages', () => {
    // Manage Lists already works; Mechanics is the same manageCategories gate.
    expect(canAccess(RoutePaths.manageLists, admin)).toBe(true);
    expect(canAccess(RoutePaths.mechanics, admin)).toBe(true);
  });
});

describe('canAccess — HR routes (top-level /hr/*, grouped under Admin)', () => {
  const hrPaths = [
    RoutePaths.hr,
    RoutePaths.hrEmployees,
    RoutePaths.hrPayroll,
    RoutePaths.hrPayslips,
    RoutePaths.hrSettings,
  ];

  it('paths are rooted at /hr', () => {
    expect(RoutePaths.hr).toBe('/hr');
    expect(RoutePaths.hrEmployees).toBe('/hr/employees');
    expect(RoutePaths.hrPayroll).toBe('/hr/payroll');
    expect(RoutePaths.hrPayslips).toBe('/hr/payslips');
    expect(RoutePaths.hrPayslipDetail).toBe('/hr/payslips/:id');
    expect(RoutePaths.hrSettings).toBe('/hr/settings');
  });

  it('admin can reach all HR pages, including a concrete payslip detail', () => {
    hrPaths.forEach((path) => {
      expect(canAccess(path, admin)).toBe(true);
    });
    expect(canAccess(`${RoutePaths.hrPayslips}/abc123`, admin)).toBe(true);
  });

  it('cashier is denied all HR pages, including a concrete payslip detail', () => {
    hrPaths.forEach((path) => {
      expect(canAccess(path, cashier)).toBe(false);
    });
    expect(canAccess(`${RoutePaths.hrPayslips}/abc123`, cashier)).toBe(false);
  });

  it('staff is denied all HR pages, including a concrete payslip detail', () => {
    hrPaths.forEach((path) => {
      expect(canAccess(path, staff)).toBe(false);
    });
    expect(canAccess(`${RoutePaths.hrPayslips}/abc123`, staff)).toBe(false);
  });

  it('old /settings/hr/* paths are not guarded routes — redirects handle them', () => {
    expect(canAccess('/settings/hr/employees', admin)).toBe(false);
    expect(canAccess('/settings/hr/payroll', admin)).toBe(false);
    expect(canAccess('/settings/hr/payslips', admin)).toBe(false);
    expect(canAccess('/settings/hr/payslips/abc123', admin)).toBe(false);
    expect(canAccess('/settings/hr/config', admin)).toBe(false);
  });
});

describe('canAccess — forgot-password', () => {
  it('is public: reachable signed out and by any role', () => {
    expect(canAccess(RoutePaths.forgotPassword, null)).toBe(true);
    expect(canAccess(RoutePaths.forgotPassword, admin)).toBe(true);
    expect(canAccess(RoutePaths.forgotPassword, cashier)).toBe(true);
  });
});

describe('canAccess — Job Orders (renamed from Drafts)', () => {
  it('paths are rooted at /job-orders', () => {
    expect(RoutePaths.jobOrders).toBe('/job-orders');
    expect(RoutePaths.jobOrderEdit).toBe('/job-orders/:id');
  });

  it('is a common route for every signed-in role, including a concrete job order', () => {
    expect(canAccess(RoutePaths.jobOrders, admin)).toBe(true);
    expect(canAccess(RoutePaths.jobOrders, cashier)).toBe(true);
    expect(canAccess(RoutePaths.jobOrders, staff)).toBe(true);
    expect(canAccess(`${RoutePaths.jobOrders}/abc123`, admin)).toBe(true);
    expect(canAccess(`${RoutePaths.jobOrders}/abc123`, cashier)).toBe(true);
    expect(canAccess(`${RoutePaths.jobOrders}/abc123`, staff)).toBe(true);
  });

  it('old /drafts paths are no longer routed — they moved, they were not aliased', () => {
    expect(canAccess('/drafts', admin)).toBe(false);
    expect(canAccess('/drafts/abc123', admin)).toBe(false);
  });
});

describe('canAccess — product edit moved into the drawer', () => {
  // The edit URL changed from /inventory/edit/:id to /inventory/:id/edit when
  // editing moved into the product drawer. The new shape matches NEITHER the
  // old '/inventory/edit/' prefix nor the single-segment '/inventory/:id'
  // view rule, so without its own rule it falls through and the gate is wrong.
  it('lets a role that may edit products reach the drawer edit route', () => {
    expect(canAccess('/inventory/p9/edit', admin)).toBe(true);
    expect(canAccess('/inventory/p9/edit', staff)).toBe(true);
  });

  it('lets a cashier into the drawer edit route (name-only editing)', () => {
    expect(canAccess('/inventory/p9/edit', cashier)).toBe(true);
  });

  it('still gates the legacy /inventory/edit/:id URL the same way', () => {
    // The redirect runs inside the router, so the guard sees the old path
    // first — it must gate identically to the new shape.
    expect(canAccess('/inventory/edit/p9', admin)).toBe(true);
    expect(canAccess('/inventory/edit/p9', cashier)).toBe(true);
  });

  it('still lets a view-only role open the product drawer', () => {
    expect(canAccess('/inventory/p9', cashier)).toBe(true);
  });
});

// Pinned when cashiers gained web access (2026-08-24) — the full mobile-parity
// access map for the cashier role.
describe('canAccess — cashier web access map', () => {
  it('reaches the mobile-parity destinations', () => {
    for (const path of [
      '/',
      '/pos',
      '/pos/checkout',
      '/job-orders',
      '/job-orders/j1',
      '/inventory',
      '/inventory/p9',
      '/expenses',
      '/expenses/add',
      '/expenses/edit/e1',
      '/reports',
      '/reports/sales',
      '/reports/labor',
      '/reports/sale/s1',
      '/sales/day',
      '/settings',
      '/settings/about',
      '/settings/lists',
      '/settings/mechanics',
    ]) {
      expect(canAccess(path, cashier)).toBe(true);
    }
  });

  it('stays out of everything cost-, admin-, or stock-facing', () => {
    for (const path of [
      '/reports/profit',
      '/reports/price-changes',
      '/inventory/add',
      '/inventory/price-history',
      // Reorder became the buying list; both views show cost.
      '/purchase-orders',
      '/purchase-orders/new',
      '/users',
      '/logs',
      '/hr',
      '/hr/payroll',
      '/receiving',
      '/receiving/new',
      '/suppliers',
      '/settings/cost-codes',
    ]) {
      expect(canAccess(path, cashier)).toBe(false);
    }
  });
});

describe('/settings/timezone', () => {
  it('is reachable by an admin', () => {
    expect(canAccess(RoutePaths.timezoneSettings, admin)).toBe(true);
  });

  // Gated on viewSettings: everyone who can open Settings can SEE the shop
  // clock (it explains why their reports roll over when they do). Editing is
  // admin-only, enforced on the page itself — same split as the mobile screen.
  it('is readable by a cashier and by staff', () => {
    expect(canAccess(RoutePaths.timezoneSettings, cashier)).toBe(true);
    expect(canAccess(RoutePaths.timezoneSettings, staff)).toBe(true);
  });

  it('is blocked for a deactivated admin', () => {
    expect(canAccess(RoutePaths.timezoneSettings, { ...admin, isActive: false })).toBe(false);
  });
});

describe('canAccess — void requests queue', () => {
  it('admins reach it', () => {
    expect(canAccess(RoutePaths.voidRequests, admin)).toBe(true);
  });

  it('cashiers do not — they file requests, they do not approve them', () => {
    // Approving voids the sale and restores stock, so the queue is gated on
    // voidSale rather than on a weaker view right.
    expect(canAccess(RoutePaths.voidRequests, cashier)).toBe(false);
    expect(canAccess(RoutePaths.voidRequests, staff)).toBe(false);
  });
});
